### PR TITLE
chore: migration tests for Auth & Custom Stacks

### PR DESCRIPTION
**Pull request by awsluja**
#### Description of changes
This PR adds migration tests for Auth, and Custom CDK stacks:

- basic auth + overrides
- complex auth
- custom cdk stack (full CX migration experience)

E2e: https://app.circleci.com/pipelines/github/aws-amplify/amplify-cli/13799/workflows/11b3bdd4-f8a2-494e-8784-1a166ba48eaf

All tests passing except for the flaky analytics windows test

#### Checklist

- [ ] PR description included
- [ ] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies
- [ ] [Pull request labels](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#labels) are added

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-migration-tests/custom-resources/custom-cdk-stack-v10.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    super(scope, id, props);
+
+    /* Do not remove - Amplify CLI automatically injects the current deployment environment in this input paramater */
+    new cdk.CfnParameter(this, 'env', {
+      type: 'String',
+      description: 'Current Amplify CLI env name',
+    });
+
+    /* AWS CDK Code goes here - Learn more: https://docs.aws.amazon.com/cdk/latest/guide/home.html */
+
```
Comment (awsluja) [inline packages/amplify-migration-tests/custom-resources/custom-cdk-stack-v10.ts:18]: This file is a sample custom resource implementation that is V10 compatible. Its contents are used in testing to simulate what the customer would create in their own project.

File: packages/amplify-migration-tests/custom-resources/custom-cdk-stack-vLatest.package.json [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    "description": "",
+    "scripts": {
+        "build": "tsc",
+        "watch": "tsc -w",
+        "test": "echo \"Error: no test specified\" && exit 1"
+    },
+    "dependencies": {
+        "@aws-amplify/cli-extensibility-helper": "^3.0.0-beta.0",
+        "aws-cdk-lib": "~2.53.0",
+        "constructs": "^10.0.5"
```
Comment (awsluja) [inline packages/amplify-migration-tests/custom-resources/custom-cdk-stack-vLatest.package.json:13]: Halfway into the migration test for custom stacks, we simulate the customer migrating from v10 to vLatest.
In doing so, they must also update the package.json file to use CDKv2 dependencies.
The contents of this file are used to simulate the effective change that customers would perform during the migration.
Notably, migrating to aws-cdk-lib, and updating the extensibility helper version.

File: packages/amplify-migration-tests/custom-resources/custom-cdk-stack-vLatest.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
+import * as cdk from 'aws-cdk-lib';
+import * as iam from 'aws-cdk-lib/aws-iam';
+import * as sns from 'aws-cdk-lib/aws-sns';
+import * as subs from 'aws-cdk-lib/aws-sns-subscriptions';
+import * as sqs from 'aws-cdk-lib/aws-sqs';
+import { Construct } from 'constructs';
+
```
Comment (awsluja) [inline packages/amplify-migration-tests/custom-resources/custom-cdk-stack-vLatest.ts:7]: We use the contents of this file to simulate the customer updating their custom stack to the vLatest (cdkv2) format.

File: packages/amplify-migration-tests/src/__tests__/migration_tests_v10/__snapshots__/auth-add-all.migration.test.ts.snap [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+Outputs
+[-] Output GoogleWebClient: {\\"Value\\":{\\"Ref\\":\\"googleClientId\\"}}
+[-] Output FacebookWebClient: {\\"Value\\":{\\"Ref\\":\\"facebookAppId\\"}}
+[-] Output AmazonWebClient: {\\"Value\\":{\\"Ref\\":\\"amazonAppId\\"}}
+[-] Output AppleWebClient: {\\"Value\\":{\\"Ref\\":\\"appleAppId\\"}}
+
+Resources
+[-] AWS::ApiGateway::Deployment DeploymentAPIGWAdminQueries97c92c5d destroy
+[+] AWS::ApiGateway::Deployment DeploymentAPIGWAdminQueries935b745b
```
Comment (awsluja) [inline packages/amplify-migration-tests/src/__tests__/migration_tests_v10/__snapshots__/auth-add-all.migration.test.ts.snap:42]: this snapshot is suspicious, could be a bug ? why is the first resource marked for destroy, and a new one being created

Comment (awsluja) [reply packages/amplify-migration-tests/src/__tests__/migration_tests_v10/__snapshots__/auth-add-all.migration.test.ts.snap:42]: this test will continue to fail until the bug is fixed in API repo

Comment (sobolk) [reply packages/amplify-migration-tests/src/__tests__/migration_tests_v10/__snapshots__/auth-add-all.migration.test.ts.snap:42]: is it possible to exclude this resource from comparison and create work item to address this later ?

**Review by sobolk — COMMENTED** [review]
LGTM.
Do we know how soon that api gateway problem could be addressed? If that timeline is unknown or long we should find a way to suppress that resource from diff and make test passing.

File: packages/amplify-migration-tests/src/__tests__/migration_tests_v10/auth-add-all.migration.test.ts [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+import { versionCheck, allowedVersionsToMigrateFrom } from '../../migration-helpers';
+import { initJSProjectWithProfileV10 } from '../../migration-helpers-v10/init';
+import { assertNoParameterChangesBetweenProjects, collectCloudformationDiffBetweenProjects, ExcludeFromCFNDiff } from '../../migration-helpers/utils';
+
+/**
+ * Due to a known limitation of APIGateway, we do not want the AWS::ApiGateway::Deployment resource
+ * from being compared in the CFN diff because it is regenerated whenever an amplify app is pulled down.
+ * This would produce a false positive when checking for differences in the CFN templates of project1 & project2.
+ * https://github.com/aws/aws-cdk/issues/8646#issuecomment-647561856
+ */
```
Comment (awsluja) [inline packages/amplify-migration-tests/src/__tests__/migration_tests_v10/auth-add-all.migration.test.ts:22]: Below is an example of how the resource changes even though the projects are identical
Screen Shot 2022-12-21 at 8 06 07 PM

Comment (awsluja) [reply packages/amplify-migration-tests/src/__tests__/migration_tests_v10/auth-add-all.migration.test.ts:22]: The code (below) deletes this property from the resource's CFN template before we do the comparison.

**Review by sobolk — APPROVED** [review]

File: packages/amplify-migration-tests/src/__tests__/migration_tests_v10/auth-add-all.migration.test.ts [hunk 6]
```diff
@@ -78,14 +78,18 @@ describe('amplify migration test auth', () => {
...
+        // on a newly pulled down project. Currently, those parameters don't get pulled down.
+
+        // await amplifyPushWithoutCodegen(projRoot2, true);
+        // assertNoParameterChangesBetweenProjects(projRoot1, projRoot2);
+        // expect(collectCloudformationDiffBetweenProjects(projRoot1, projRoot2, cfnDiffExclusions)).toMatchSnapshot();
+
+        // // should be able to remove & add auth after pulling down an older project
+        // await removeAuthWithDefault(projRoot2, true);
+        // await addAuthWithDefault(projRoot2, {}, true);
+        // await amplifyPushAuth(projRoot2, true);
```
Comment (sobolk) [inline packages/amplify-migration-tests/src/__tests__/migration_tests_v10/auth-add-all.migration.test.ts:92]: I'd rather create follow up work item for this than leaving commented out code.

**Review by sobolk — APPROVED** [review]

**Review by akshbhu — APPROVED** [review]

**Pull request merged**